File: app/src/main/java/com/cyl/musiclake/api/netease/NeteaseApiServiceImpl.java
package com.cyl.musiclake.api.netease;

import com.cyl.musiclake.data.db.Music;
import com.cyl.musiclake.data.db.Playlist;
import com.cyl.musiclake.common.Constants;
import com.cyl.musiclake.net.ApiManager;
import com.cyl.musiclake.utils.LogUtil;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.Observable;
import io.reactivex.ObservableOnSubscribe;

/**
 * Created by D22434 on 2018/1/5.
 */

public class NeteaseApiServiceImpl {
    private static final String TAG = "NeteaseApiServiceImpl";

    public static NeteaseApiService getApiService() {
        return ApiManager.getInstance().create(NeteaseApiService.class, Constants.BASE_NETEASE_URL);
    }

    public static Observable<Playlist> getTopList(int ids) {
        return getApiService().getTopList(ids)
                .flatMap(topList -> Observable.create((ObservableOnSubscribe<Playlist>) e -> {
                    try {
                        Playlist playlist = new Playlist();
                        playlist.setPid(String.valueOf(topList.getResult().getId()));
                        playlist.setName(topList.getResult().getName());
                        playlist.setDate(topList.getResult().getTrackUpdateTime());
                        playlist.setDes(topList.getResult().getDescription());
                        List<Music> musicList = new ArrayList<>();
                        for (NeteaseMusic neteaseMusic : topList.getResult().getTracks()) {
                        }
                        playlist.setCoverUrl(topList.getResult().getCoverImgUrl());
                        playlist.setMusicList(musicList);
                        e.onNext(playlist);
                        e.onComplete();
                    } catch (Exception ep) {
                        e.onError(ep);
                    }
                }));
    }

    public static Observable<List<NeteaseList>> getTopPlaylist() {
        return getApiService().getNeteasePlaylist()
                .flatMap(topList -> Observable.create((ObservableOnSubscribe<List<NeteaseList>>) e -> {
                    try {
                        e.onNext(topList.getPlaylists());
                        e.onComplete();
                    } catch (Exception ep) {
                        e.onError(ep);
                    }
                }));
    }

    public static Observable<Music> getMusicUrl(Music music) {
        return getApiService().getMusicUrl(music.getMid())
                .flatMap(neteaseMusicUrl -> Observable.create((ObservableOnSubscribe<Music>) e -> {
                    try {
                        if (neteaseMusicUrl.getCode() == 200) {
                            music.setUri(neteaseMusicUrl.getData().get(0).getUrl());
                        }
                        e.onNext(music);
                        e.onComplete();
                    } catch (Exception ep) {
                        e.onError(ep);
                    }
                }));
    }

    public static Observable<List<Music>> search(String key, int limit, int page) {
        Map<String, String> params = new HashMap<>();
        params.put("offset", String.valueOf(page)); //page
        params.put("limit", String.valueOf(limit));//limit
        params.put("keywords", key);// key
        return getApiService().searchByNetease(params)
                .flatMap(netease -> {
                    List<Music> musicList = new ArrayList<>();
                    List<NeteaseMusic> songList = netease.getResult().getSongs();
                    for (int i = 0; i < songList.size(); i++) {
                    }
                    LogUtil.e("search", page + "--" + limit + "qq :" + musicList.size());
                    return Observable.fromArray(musicList);
                });
    }

    public static Observable<String> getNeteaseLyric(Music music) {
        return getApiService().getMusicLyric(music.getMid())
                .flatMap(netease -> {
                    String lyric = netease.getLrc().getLyric();
                    LogUtil.e("lyric =", lyric);
                    return Observable.fromArray(lyric);
                });
    }
}

File: app/src/main/java/com/cyl/musiclake/api/netease/NeteaseApiServiceImpl.kt
package com.cyl.musiclake.api.netease

import com.cyl.musicapi.netease.*
import com.cyl.musiclake.bean.Artist
import com.cyl.musiclake.bean.HotSearchBean
import com.cyl.musiclake.common.Constants
import com.cyl.musiclake.net.ApiManager
import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe

/**
 * Created by D22434 on 2018/1/5.
 */

object NeteaseApiServiceImpl {
    private val TAG = "NeteaseApiServiceImpl"

    val apiService by lazy { ApiManager.getInstance().create(NeteaseApiService::class.java, Constants.BASE_NETEASE_URL) }

    /**
     * 获取歌单歌曲
     */
    fun getTopArtists(limit: Int, offset: Int): Observable<MutableList<Artist>> {
        return apiService.getTopArtists(offset, limit)
                .flatMap { it ->
                    Observable.create(ObservableOnSubscribe<MutableList<Artist>> { e ->
                        try {
                            if (it.code == 200) {
                                val list = mutableListOf<Artist>()
                                it.list.artists?.forEach {
                                    val artistsBean = Artist()
                                    artistsBean.id = it.id.toLong()
                                    artistsBean.name = it.name
                                    artistsBean.picUrl = it.picUrl
                                    artistsBean.score = it.score
                                    artistsBean.musicSize = it.musicSize
                                    artistsBean.albumSize = it.albumSize
                                    artistsBean.type = Constants.NETEASE
                                    list.add(artistsBean)
                                }
                                e.onNext(list)
                                e.onComplete()
                            } else {
                                e.onError(Throwable("网络异常"))
                            }
                        } catch (ep: Exception) {
                            e.onError(ep)
                        }
                    })
                }
    }

    /**
     * 获取推荐mv
     */
    fun getNewestMv(limit: Int): Observable<MvInfo> {
        return apiService.getNewestMv(limit)
    }

    /**
     * 获取推荐mv
     */
    fun getTopMv(limit: Int, offset: Int): Observable<MvInfo> {
        return apiService.getTopMv(offset, limit)
    }

    /**
     * 获取mv信息
     */
    fun getMvDetailInfo(mvid: String): Observable<MvDetailInfo> {
        return apiService.getMvDetailInfo(mvid)
    }

    /**
     * 获取相似mv
     */
    fun getSimilarMv(mvid: String): Observable<SimilarMvInfo> {
        return apiService.getSimilarMv(mvid)
    }

    /**
     * 获取mv评论
     */
    fun getMvComment(mvid: String): Observable<MvComment> {
        return apiService.getMvComment(mvid)
    }

    /**
     * 获取mv评论
     */
    fun getHotSearchInfo(): Observable<MutableList<HotSearchBean>> {
        return apiService.getHotSearchInfo()
                .flatMap { it ->
                    Observable.create(ObservableOnSubscribe<MutableList<HotSearchBean>> { e ->
                        try {
                            if (it.code == 200) {
                                val list = mutableListOf<HotSearchBean>()
                                it.result.hots?.forEach {
                                    list.add(HotSearchBean(it.first))
                                }
                                e.onNext(list)
                                e.onComplete()
                            } else {
                                e.onError(Throwable("网络异常"))
                            }
                        } catch (ep: Exception) {
                            e.onError(ep)
                        }
                    })
                }
    }
}
